lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,390 @@
+  SmallVector<Attribute> brgemmFlags;
+  if (info.isInitOutput) {
+    brgemmFlags.push_back(microkernel::BrgemmFlagsAttr::get(
+        rewriter.getContext(), microkernel::BrgemmFlags::BETA_0));
+  }
+  if (info.mode == BrgemmInfo::STRIDE_MODE) {
+    brgemmFlags.push_back(microkernel::BrgemmFlagsAttr::get(
+        rewriter.getContext(), microkernel::BrgemmFlags::STRIDE));
+  } else if (info.mode == BrgemmInfo::LIST_MODE) {
+    brgemmFlags.push_back(microkernel::BrgemmFlagsAttr::get(
+        rewriter.getContext(), microkernel::BrgemmFlags::LIST));
+  }
+
+  SmallVector<Attribute, 2> brgemmDtypes{
+      TypeAttr::get(getElementTypeOrSelf(linalgOp.getDpsInputs()[0].getType())),
+      TypeAttr::get(
+          getElementTypeOrSelf(linalgOp.getDpsInputs()[1].getType()))};
+
+  // create dispatch op
+  auto flags = rewriter.getArrayAttr(brgemmFlags);
+  auto dtypes = rewriter.getArrayAttr(brgemmDtypes);
+  DenseI64ArrayAttr dims = DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{info.m, info.n, info.k, info.lda, info.ldb, info.ldc,
+                        info.strideA, info.strideB});
+  Value dispatched = rewriter.create<microkernel::BrgemmDispatchOp>(
+      loc, integer64, dims, flags, dtypes);
+
+  // create prologue op
+  rewriter.create<microkernel::BrgemmPrologueOp>(loc, dispatched);
+
+  // create brgemm invoke op
+  Value batchDim = rewriter.create<arith::ConstantOp>(
+      loc, integer64, rewriter.getIntegerAttr(integer64, info.batchSize));
+  Value lenDim = rewriter.create<arith::ConstantOp>(
+      loc, integer64, rewriter.getIntegerAttr(integer64, info.addrLen));
+  SmallVector<Value> invokeOperands;
+  invokeOperands.push_back(dispatched);
+  invokeOperands.append(linalgOp->getOperands().begin(),
+                        linalgOp->getOperands().end());
+  invokeOperands.push_back(batchDim);
+  invokeOperands.push_back(lenDim);
+  rewriter.create<microkernel::BrgemmOp>(loc, invokeOperands);
+
+  // create epilogue op & replace original op
+  rewriter.replaceOpWithNewOp<microkernel::BrgemmEpilogueOp>(linalgOp,
+                                                             dispatched);
+}
+
+bool isZeroArithConstant(arith::ConstantOp op) {
+  if (!op)
+    return false;
+
+  if (auto intAttr = llvm::dyn_cast<IntegerAttr>(op.getValue())) {
+    if (intAttr.getInt() != 0)
+      return false;
+  } else if (auto floatAttr = llvm::dyn_cast<FloatAttr>(op.getValue())) {
+    if (!floatAttr.getValue().isZero())
+      return false;
+  } else
+    return false;
+
+  return true;
+}
+
+template <typename ContractionOp>
+class ConvertContractionOpToBrgemmRewriter
+    : public OpRewritePattern<ContractionOp> {
+public:
+  using OpRewritePattern<ContractionOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ContractionOp op,
+                                PatternRewriter &rewriter) const final {
+    auto brgemmInfo = getBrgemmInfo(op);
+    if (failed(brgemmInfo))
+      return failure();
```

Comment:
Can this be an assertion or is there a valid situation when the info generation can fail?